worlds/earthbound/Client.py

```diff
@@ -0,0 +1,510 @@
+        return True
+
+    async def game_watcher(self, ctx: "SNIContext") -> None:
+        from SNIClient import snes_buffered_write, snes_flush_writes, snes_read, snes_write
+        giygas_clear = await snes_read(ctx, GIYGAS_CLEAR, 0x1)
+        game_clear = await snes_read(ctx, GAME_CLEAR, 0x1)
+        item_received = await snes_read(ctx, ITEM_RECEIVED, 0x1)
+        special_received = await snes_read(ctx, SPECIAL_RECEIVED, 0x1)
+        money_received = await snes_read(ctx, MONEY_RECIVED, 0x2)
+        save_num = await snes_read(ctx, SAVE_FILE, 0x1)
+        text_open = await snes_read(ctx, OPEN_WINDOW, 1)
+        melody_table = await snes_read(ctx, MELODY_TABLE, 2)
+        earth_power_absorbed = await snes_read(ctx, EARTH_POWER_FLAG, 1)
+        cur_script = await snes_read(ctx, CUR_SCENE, 1)
+        rom = await snes_read(ctx, EB_ROMHASH_START, ROMHASH_SIZE)
+        scouted_hint_flags = await snes_read(ctx, SCOUTED_HINT_FLAGS, 1)
+        gift_target = await snes_read(ctx, WRAM_START + 0xB5E7, 2)
+        outbound_gifts = await snes_read(ctx, WRAM_START + 0x31D0, 1)
+        shop_scout = await snes_read(ctx, WRAM_START + 0x0770, 1)
+        shop_scouts_enabled = await snes_read(ctx, ROM_START + 0x04FD77, 1)
+        outgoing_energy = await snes_read(ctx, MONEY_IN_BANK, 4)
+        if rom != ctx.rom:
+            ctx.rom = None
+            return
+        
+        if giygas_clear[0] & 0x01 == 0x01:  # Are we in the epilogue
+            return
+
+        if save_num[0] == 0x00:  # If on the title screen
+            return
+
+        if ctx.slot is None:
+            return
+
+        if outgoing_energy is None: #None Catcher
+            return
+
+        await ctx.send_msgs([{
+            "cmd": "Get",
+            "keys": [f"GiftBoxes;{ctx.team}"]
+        }])
+
+        # GIFTING DATA
+        if f"GiftBox;{ctx.team};{ctx.slot}" not in ctx.stored_data:
+            local_giftbox = {
+                            str(ctx.slot): {
+                                "is_open": True,
+                                "accepts_any_gift": False,
+                                "desired_traits": wanted_traits,
+                                "minimum_gift_data_version": 2,
+                                "maximum_gift_data_version": 3}}
+            await ctx.send_msgs([{
+                        "cmd": "Set",
+                        "key": f"GiftBoxes;{ctx.team}",
+                        "want_reply": False,
+                        "default": {},
+                        "operations": [{"operation": "update", "value": local_giftbox}]
+                    }])
+        
+            await ctx.send_msgs([{
+                        "cmd": "Get",
+                        "keys": [f"GiftBox;{ctx.team};{ctx.slot}"]
+                    }])
+
+        await ctx.send_msgs([{
+                    "cmd": "SetNotify",
+                    "keys": [f"GiftBox;{ctx.team};{ctx.slot}", f"GiftBoxes;{ctx.team}"]
+                }])
```

Comment:
Re my last comment: This code block helps understand what I mean.

The first part of this is really good. You only send the initial "Set" and "Get" if the Gift Box is not already initialized.
But the SetNotify is outside of this `if`. But why? It also only has to happen once. Instead, this SetNotify gets sent over and over again, multiple times per second.